cio/io_test.go

```diff
@@ -87,7 +87,6 @@ func TestNewAttach(t *testing.T) {
 	actualStdin, err := ioutil.ReadAll(producers.Stdin)
 	require.NoError(t, err)
 
-	io.Cancel()
```

Comment:
Why is this change necessary?

This `Cancel()` is here because every use of `cio.IO` calls cancel before `Close()`. If we don't include it in the test, the test isn't really correct.

Reply:
I removed it because the tests just failed on a race in that test where stderr and stdout had "" results.

Reply:
I don't think you should call cancel before Wait.  Cancel should only be needed if the callers are not closing the pipes.  If you close them all, you shouldn't have to use cancel

Reply:
Swapping the order of `Cancel()` and `Wait()` also seems to fix the test error, and in this way it also keeps the test rule of calling `Cancel()` before `Close()`.

Reply:
`Cancel()` shouldn't need to be called at all.  Its only to cancel/abort operations

Reply:
Besides, why would we ever make an API where you HAD to call Cancel before Close?  I makes no sense at all and is horrible design.  Just put the cancel in the Close() call.  It's two separate methods for a reason.  It makes no sense and is ugly and a dumb API if you HAVE to call cancel before Close.

Reply:
The caller is `containerd`. The "user" implements `IO`, and passed it to `containerd`. The user doesn't call the `IO` methods, so as a user of `IO` why should I have to call `io.Cancel()` in every `io.Close()`. That's a bad interface.

Reply:
I really don't understand.  Cancel should only be called when you want to cancel io operations or else you get pipes that are not properly flushed and the race and output here are the types of results that you get because of this.

Reply:
Technically, calling `Cancel()` in the `Close()` is only done once by the implementer, it's called automatically afterwards.

I think I said so in a previous PR when we were discussing this, but I don't mind either way. Usually though, closing a thing cancel all other related operations (take a FD for instance). Closing something tell that the intent is to stop using it, so if that requires a Cancel it doesn't shock me.

For `OpenFifos()`, `Close()` will actually have the same effect as cancelling the context normally btw.

Reply:
> Cancel should only be called when you want to cancel io operations 

My point is that the `IO` interface is not something that `containerd` provides to consumers. It's an "input" interface. `containerd` accepts `IO` it never returns an `IO`. With such an interface, the actor calling the methods is not the consumer of `contaienrd`, the actor is `containerd`. So the "you" in the above statement is an implementation detail of `containerd`. The consumer only needs to be concerned about "how do I correctly implement an `IO`".  It's easier to document and understand when we say `Cancel()` will always be called before `Close()`.

That said, in this test case it doesn't matter because the implementation works just fine if `Close()` is called without `Cancel()`. If we want to force all implementations to behave this way then we need to revert the docs change in #1894 and update it to say that `Close()` must `Cancel()` itself first.